File: liu/v.39/sending_mail.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <limits>
#define MVI unordered_map<int, vector<int[2]>>

using namespace std;

string dijkstras(int n, int m, int s, int t, MVI nodes, int c){
    if (m == 0) return "Case #" + to_string(c) + ": unreachable";
    int distance[n];
    for (int i = 0; i < n; i++) distance[i] = numeric_limits<int>::max();
    distance[s] = 0;
    vector<int> prioQueue;
    prioQueue.push_back(s);
    while (!prioQueue.empty()){
        int min = numeric_limits<int>::max();
        int cn, index;
        for (int i = 0; i < prioQueue.size(); i++){
            if (distance[prioQueue[i]] < min){
                cn = prioQueue[i];
                index = i;
                min = distance[prioQueue[i]];
            }
        }
        prioQueue.erase(prioQueue.begin()+index);
        if (cn == t) return "Case #" + to_string(c) + ": " + to_string(distance[cn]);
        for (int i = 0; i < nodes[cn]; i++)
        {
            /* code */
        }
        
    }
    return "Case #" + to_string(c) + ": unreachable";
}

int main(){
    int N, n, m, s, t, n1, n2, w;
    cin >> N;
    for (int c = 0; c < N; c++)
    {
        cin >> n >> m >> s >> t;
        MVI nodes;
        for (int i = 0; i < n; i++) nodes[i] = vector<int[2]>;
        
        for (int i = 0; i < m; i++)
        {
            cin >> n1 >> n2 >> w;
            nodes[n1].push_back({n2, w});
            nodes[n2].push_back({n1, w});
        }
        cout << dijkstras(n, m, s, t, nodes, c) << endl;
    }
    return 0;
}

File: liu/v.39/sending_mail.py
N = int(input())

def dijkstras(n, m, s, t, nodes, i):
    distance = [float('inf') for _ in range(n)]
    distance[s] = 0
    prio_queue = [s]
    while True:
        if not prio_queue or m == 0:
            print("Case #" + str(i) + ": unreachable")
            break
        # Take out smallest distance in prio_queue
        min = float('inf')
        cn = None
        for node in prio_queue:
            if distance[node] < min:
                cn = node
                min = distance[node]
        prio_queue.remove(cn)
        # Check if we have arrived
        if cn == t:
            print("Case #" + str(i) + ": " + str(distance[cn]))
            break
        # Search through after neighbours
        for nb_w in nodes[cn]:
            if distance[cn] + nb_w[1] < distance[nb_w[0]]: 
                distance[nb_w[0]] = distance[cn] + nb_w[1]
                if nb_w[0] not in prio_queue:
                    prio_queue.append(nb_w[0])

for i in range(1, N+1):
    n, m, s, t = map(int, input().split(" "))
    nodes = {}
    for j in range(n): nodes[j] = []
    for j in range(m):
        n1, n2, w = map(int, input().split(" "))
        nodes[n1].append((n2, w))
        nodes[n2].append((n1, w))
    
    dijkstras(n, m, s, t, nodes, i)
    